Spaces Swift SDK example app

andrewjl-mux opened the pull request:
- add Spaces Swift SDK example app
- update .gitignore exclusions

refactornator commented on README.md:
```diff
@@ -17,6 +17,7 @@ Get to know Mux by looking through examples! Contributions are always welcome, e
 | [swift-video-app](swift-video-app)                      | Swift app that uses Mux Video, Mux Data and plays audio when the app goes into the background | VOD, Mobile | Swift |
 | [aws-recommendation-engine](aws-recommendation-engine)  | An example of using Mux Data streaming exports to recommend content with Amazon Personalize | VOD, Serverless | TypeScript |
 | [nextjs-mdx-player](nextjs-mdx-player)                  | A [Next.js](https://nextjs.org/) app configured to support [MDX](https://mdxjs.com) to illustrate supporting [MuxPlayer](https://www.mux.com/player) in blog posts. | VOD, CMS | JavaScript |
+| [swift-spaces-real-time-app](swift-spaces-real-time-app)| An example iOS and iPadOS app demonstrating usage of the Spaces Swift SDk for real-time video
```
Probably good to add tags and language to this.

refactornator commented on swift-spaces-real-time-app/README.md:
```diff
@@ -0,0 +1,23 @@
+
+### Adding the SDK manually
+
+Follow the steps here to add the SDK manually.
```
It's unclear which steps to follow here.

andrewjl-mux replied:
I've removed this section.

refactornator commented on swift-spaces-real-time-app/MuxSpacesUIKit/Space+Token.swift:
```diff
@@ -0,0 +1,23 @@
+extension Space {
+    static func make() throws -> Space {
+        if token.isEmpty {
+            print("Please add a include Spaces JWT to try out the sample code")
```
```suggestion
            print("Please include a Spaces JWT to try out the sample code")
```

refactornator commented on swift-spaces-real-time-app/MuxSpacesUIKit/View/JoinSpaceViewController.swift:
```diff
@@ -0,0 +1,548 @@
+
+    var cancellables: [AnyCancellable] = []
+
+    @Published var isAudioPublishingEnabled: Bool = false
+    @Published var isVideoPublishingEnabled: Bool = false
```
```suggestion
    @Published var isAudioPublishingEnabled: Bool = true
    @Published var isVideoPublishingEnabled: Bool = true
```

refactornator replied:
I think it makes more sense to default this to enabled so audio and video are published on joining.

refactornator commented on swift-spaces-real-time-app/MuxSpacesUIKit/View/JoinSpaceViewController.swift:
```diff
@@ -0,0 +1,548 @@
+        ).withSize(16.0)
+        audioLabel.textColor = .label
+        audioToggle.translatesAutoresizingMaskIntoConstraints = false
+        audioToggle.isEnabled = true
```
```suggestion
        audioToggle.isEnabled = true
        audioToggle.isOn = true
```

refactornator commented on swift-spaces-real-time-app/MuxSpacesUIKit/View/JoinSpaceViewController.swift:
```diff
@@ -0,0 +1,548 @@
+        ).withSize(16.0)
+        videoLabel.textColor = .label
+        videoToggle.translatesAutoresizingMaskIntoConstraints = false
+        videoToggle.isEnabled = true
```
```suggestion
        videoToggle.isEnabled = true
        videoToggle.isOn = true
```

refactornator reviewed (approved):
I got the sample app to run on my phone. I made a few suggestions like defaulting to publishing audio and video. As well as a wording fix. The sample app works, so I think we're good to merge it and we can continue to improve it.

One surprising thing was that I had to "reset package cache" for it to find the MuxSpaces dependency and not fail. I'm not sure if there's any way to prevent that from needing to be reset.

<img width="399" alt="Screen Shot 2022-10-28 at 7 09 22 PM" src="https://user-images.githubusercontent.com/1211390/198768608-4273495e-1a11-4af5-82a1-2eee086eee65.png">

andrewjl-mux commented:
> I got the sample app to run on my phone. I made a few suggestions like defaulting to publishing audio and video. As well as a wording fix. The sample app works, so I think we're good to merge it and we can continue to improve it.

Thanks for flagging these! The most recent changes improve the state handling further and these publishing audio and video flags ended up being duplicate state, so they were removed. The spaces token is now also injected as an environment variable.

> One surprising thing was that I had to "reset package cache" for it to find the MuxSpaces dependency and not fail. I'm not sure if there's any way to prevent that from needing to be reset.

Good catch, fixed this by removing the Swift package from the Xcode project. This does mean that the SDK needs to be installed first for the app to be usable. I updated the app README to highlight this.

refactornator commented:
I pulled the latest changes and added the MuxSpaces package from SPM and successfully ran the app 🎉

One odd inconsistency is that when I open the app, the toggles for publishing audio and video look off, but the app shows the audio and video options I get when they are toggled on. I think the internal state is that they are on. Maybe we should just remove the toggles and always publish audio and video? For a sample app, I would be ok with this, if you agree.
![IMG_6360](https://user-images.githubusercontent.com/1211390/199077685-235556c2-a426-415e-b5ef-4f93fea3131d.PNG)

I was able to successfully join a space and everything seemed to be working there. But I noticed that if I left the space, and immediately tried to rejoin, I got an error that my JWT was invalid. This might be annoying for anyone testing with the sample app. The JWT should still be valid and it should be possible to rejoin the space.

andrewjl-mux commented:
Both the toggles and re-joining the space have been fixed, please keep the feedback coming and I will address in a follow-on PR.
